sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Product.scala

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.expressions.aggregate
+
+import org.apache.spark.sql.catalyst.analysis.TypeCheckResult
+import org.apache.spark.sql.catalyst.dsl.expressions._
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.util.TypeUtils
+import org.apache.spark.sql.types.{ AbstractDataType, DataType, DoubleType, NumericType }
+
+
+@ExpressionDescription(
+  usage = "_FUNC_(expr) - Returns the product calculated from values of a group.",
+  examples = """
```

Comment:
@HyukjinKwon oh I probably just don't understand this part but does this not make it usable in SQL?

Reply:
Ah .. yeah. It's a bit odds that we do it with `ExpressionDescription` here. BTW, it has to be added into here too: https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala#L199-L579 to be used in SQL.

@rwpenney can you remove this annotation? Seems like it's confusing. I think it makes sense not to add in SQL side for now.

Reply:
WDYT @cloud-fan?

Reply:
I can't find this PRODUCT function in other databases, and seems it's not a useful feature for SQL users: https://stackoverflow.com/questions/3912204/why-is-there-no-product-aggregate-function-in-sql

Let's not add it for now.

Reply:
Thanks, all, for comments and advice so far. As you can probably tell, I'm very new to the internals of Spark.

I'm inclined to agree that SQL users may be less likely to value this "product" function than Scala/Python/R users, although I'd assumed that those `@ExpressionDescription` annotations would automatically make it visible in SQL. From my use-cases as a Data Scientist, these Scala/Python/R interfaces are much more appropriate for the types of statistical number-crunching where the "product" aggregation will be valuable.

I'm happy to remove those SQL-specific annotations if that's the consensus here.

Reply:
Yeah, let's remove. I think we all agree here to remove it in SQL.

Reply:
`@ExpressionDescription` now removed, so `product()` should only be visible via Scala/Python/R APIs.